### ISSUE TITLE
CallMolecularConsensusReads Error java.lang.ArrayIndexOutOfBoundsException

### ISSUE THREAD

**Issue by jst43** [fulcrumgenomics/fgbio#531]
I have a mapped BAM with UMIs on which I've already run the following commands without error: bwa mem; AnnotateBamWithUmis; SortBam; SetMateInformation; and GroupReadsByUmi.

I'm now attempting to run the following CallMolecularConsensusReadsCommand:
`fgbio --tmp-dir=./fgbio_tmp CallMolecularConsensusReads -1 41 -i ./groupby_umi.bam -o ./call_consensus.bam -M 2 2> ./call_consensus.log`

However the task keeps failing, with the log file showing the error:

    Time for last 2,000,000:  242s.  Last read position: */*
    [2019/09/18 03:17:53 | CallMolecularConsensusReads | Info] processed    12,500,000 records.  Elapsed time: 00:04:05s.  Time for last 500,000:   12s.  Last read position: chr8:143,563,991
    [2019/09/18 03:18:03 | CallMolecularConsensusReads | Info] processed    13,000,000 records.  Elapsed time: 00:04:16s.  Time for last 500,000:   10s.  Last read position: chr9:99,386,947
    [2019/09/18 03:18:04 | FgBioMain | Info] CallMolecularConsensusReads failed. Elapsed time: 4.32 minutes.
    Exception in thread "main" java.lang.ArrayIndexOutOfBoundsException: 0
        at com.fulcrumgenomics.umi.UmiConsensusCaller.$anonfun$toSourceRead$1(UmiConsensusCaller.scala:210)
        at com.fulcrumgenomics.commons.CommonsDef.forloop(CommonsDef.scala:210)
        at com.fulcrumgenomics.umi.UmiConsensusCaller.toSourceRead(UmiConsensusCaller.scala:209)
        at com.fulcrumgenomics.umi.UmiConsensusCaller.toSourceRead$(UmiConsensusCaller.scala:195)
        at com.fulcrumgenomics.umi.VanillaUmiConsensusCaller.toSourceRead(VanillaUmiConsensusCaller.scala:102)
        at com.fulcrumgenomics.umi.VanillaUmiConsensusCaller.$anonfun$consensusFromSamRecords$1(VanillaUmiConsensusCaller.scala:152)
        at scala.collection.immutable.Stream.flatMap(Stream.scala:486)
        at scala.collection.immutable.Stream.$anonfun$flatMap$1(Stream.scala:494)
        at scala.collection.immutable.Stream.$anonfun$append$1(Stream.scala:252)
        at scala.collection.immutable.Stream$Cons.tail(Stream.scala:1169)
        at scala.collection.immutable.Stream$Cons.tail(Stream.scala:1159)
        at scala.collection.immutable.Stream.length(Stream.scala:309)
        at scala.collection.SeqLike.sorted(SeqLike.scala:635)
        at scala.collection.SeqLike.sorted$(SeqLike.scala:634)
        at scala.collection.AbstractSeq.sorted(Seq.scala:41)
        at scala.collection.SeqLike.sortBy(SeqLike.scala:621)
        at scala.collection.SeqLike.sortBy$(SeqLike.scala:621)
        at scala.collection.AbstractSeq.sortBy(Seq.scala:41)
        at com.fulcrumgenomics.umi.UmiConsensusCaller.filterToMostCommonAlignment(UmiConsensusCaller.scala:290)
        at com.fulcrumgenomics.umi.UmiConsensusCaller.filterToMostCommonAlignment$(UmiConsensusCaller.scala:286)
        at com.fulcrumgenomics.umi.VanillaUmiConsensusCaller.filterToMostCommonAlignment(VanillaUmiConsensusCaller.scala:102)
        at com.fulcrumgenomics.umi.VanillaUmiConsensusCaller.consensusFromSamRecords(VanillaUmiConsensusCaller.scala:153)
        at com.fulcrumgenomics.umi.VanillaUmiConsensusCaller.consensusSamRecordsFromSamRecords(VanillaUmiConsensusCaller.scala:133)
        at com.fulcrumgenomics.umi.UmiConsensusCaller.consensusReadsFromSamRecords(UmiConsensusCaller.scala:248)
        at com.fulcrumgenomics.umi.UmiConsensusCaller.consensusReadsFromSamRecords$(UmiConsensusCaller.scala:246)
        at com.fulcrumgenomics.umi.VanillaUmiConsensusCaller.consensusReadsFromSamRecords(VanillaUmiConsensusCaller.scala:102)
        at com.fulcrumgenomics.umi.ConsensusCallingIterator.advance(ConsensusCallingIterator.scala:78)
        at com.fulcrumgenomics.umi.ConsensusCallingIterator.hasNext(ConsensusCallingIterator.scala:50)
        at scala.collection.Iterator.foreach(Iterator.scala:929)
        at scala.collection.Iterator.foreach$(Iterator.scala:929)
        at com.fulcrumgenomics.umi.ConsensusCallingIterator.foreach(ConsensusCallingIterator.scala:41)
        at com.fulcrumgenomics.Writer.write(Writer.scala:37)
        at com.fulcrumgenomics.Writer.write$(Writer.scala:37)
        at com.fulcrumgenomics.bam.api.SamWriter.write(SamWriter.scala:101)
        at com.fulcrumgenomics.Writer.$plus$plus$eq(Writer.scala:47)
        at com.fulcrumgenomics.Writer.$plus$plus$eq$(Writer.scala:46)
        at com.fulcrumgenomics.bam.api.SamWriter.$plus$plus$eq(SamWriter.scala:101)
        at com.fulcrumgenomics.umi.CallMolecularConsensusReads.execute(CallMolecularConsensusReads.scala:165)
        at com.fulcrumgenomics.cmdline.FgBioMain.makeItSo(FgBioMain.scala:99)
        at com.fulcrumgenomics.cmdline.FgBioMain.makeItSoAndExit(FgBioMain.scala:80)
        at com.fulcrumgenomics.cmdline.FgBioMain$.main(FgBioMain.scala:48)
        at com.fulcrumgenomics.cmdline.FgBioMain.main(FgBioMain.scala)

Any advice on what might be causing the error would be greatly appreciated

**Comment by tfenne** [comment]
@jst43 Could you please share what version of fgbio you are using?  The latest version has a comment at `UmiConsensusCaller.scala:210` so I'm struggling a little bit to figure out exactly which line is causing the exception.

My best guess given the stack trace is that there's something funny with the bases and/or quals for a read in your BAM file.  If I had to guess, I would guess that you have a record in the BAM file where the qualities are either missing, or are somehow shorter than the bases.

**Comment by jst43** [comment]
@tfenne I originally ran it using v0.4, and then when I encountered the error I upgraded to v1.0.0 to see if that would help. I thought I'd pasted in the error message for v1.0.0 but I guess I used v0.4 by mistake. Here's the error message for v1.0.0:

    Time for last 2,000,000:  266s.  Last read position: */*
    [2019/09/19 03:25:41 | CallMolecularConsensusReads | Info] processed    12,500,000 records.  Elapsed time: 00:04:28s.  Time for last 500,000:   25s.  Last read position: chr8:143,568,803
    [2019/09/19 03:25:51 | CallMolecularConsensusReads | Info] processed    13,000,000 records.  Elapsed time: 00:04:38s.  Time for last 500,000:   10s.  Last read position: chr9:99,397,158
    [2019/09/19 03:25:51 | FgBioMain | Info] CallMolecularConsensusReads failed. Elapsed time: 4.69 minutes.
    Exception in thread "main" java.lang.ArrayIndexOutOfBoundsException: 0
	    at com.fulcrumgenomics.umi.UmiConsensusCaller.$anonfun$toSourceRead$1(UmiConsensusCaller.scala:238)
	    at com.fulcrumgenomics.commons.CommonsDef.forloop(CommonsDef.scala:305)
	    at com.fulcrumgenomics.commons.CommonsDef.forloop$(CommonsDef.scala:301)
	    at com.fulcrumgenomics.FgBioDef$.forloop(FgBioDef.scala:35)
	    at com.fulcrumgenomics.umi.UmiConsensusCaller.toSourceRead(UmiConsensusCaller.scala:237)
	    at com.fulcrumgenomics.umi.UmiConsensusCaller.toSourceRead$(UmiConsensusCaller.scala:223)
	    at com.fulcrumgenomics.umi.VanillaUmiConsensusCaller.toSourceRead(VanillaUmiConsensusCaller.scala:101)
	    at com.fulcrumgenomics.umi.VanillaUmiConsensusCaller.$anonfun$consensusFromSamRecords$1(VanillaUmiConsensusCaller.scala:160)
	    at scala.collection.StrictOptimizedIterableOps.flatMap(StrictOptimizedIterableOps.scala:118)
	    at scala.collection.StrictOptimizedIterableOps.flatMap$(StrictOptimizedIterableOps.scala:105)
	    at scala.collection.immutable.Vector.flatMap(Vector.scala:113)
	    at com.fulcrumgenomics.umi.VanillaUmiConsensusCaller.consensusFromSamRecords(VanillaUmiConsensusCaller.scala:160)
	    at com.fulcrumgenomics.umi.VanillaUmiConsensusCaller.consensusSamRecordsFromSamRecords(VanillaUmiConsensusCaller.scala:141)
	    at com.fulcrumgenomics.umi.UmiConsensusCaller.consensusReadsFromSamRecords(UmiConsensusCaller.scala:282)
	    at com.fulcrumgenomics.umi.UmiConsensusCaller.consensusReadsFromSamRecords$(UmiConsensusCaller.scala:280)
	    at com.fulcrumgenomics.umi.VanillaUmiConsensusCaller.consensusReadsFromSamRecords(VanillaUmiConsensusCaller.scala:101)
	    at com.fulcrumgenomics.umi.ConsensusCallingIterator.$anonfun$iter$2(ConsensusCallingIterator.scala:63)
	    at scala.collection.Iterator$$anon$10.current(Iterator.scala:562)
	    at scala.collection.Iterator$$anon$10.hasNext(Iterator.scala:566)
	    at com.fulcrumgenomics.umi.ConsensusCallingIterator.hasNext(ConsensusCallingIterator.scala:104)
	    at scala.collection.IterableOnceOps.foreach(IterableOnce.scala:576)
	    at scala.collection.IterableOnceOps.foreach$(IterableOnce.scala:574)
	    at com.fulcrumgenomics.umi.ConsensusCallingIterator.foreach(ConsensusCallingIterator.scala:48)
	    at com.fulcrumgenomics.commons.io.Writer.write(Writer.scala:36)
	    at com.fulcrumgenomics.commons.io.Writer.write$(Writer.scala:36)
	    at com.fulcrumgenomics.bam.api.SamWriter.write(SamWriter.scala:101)
	    at com.fulcrumgenomics.commons.io.Writer.$plus$plus$eq(Writer.scala:46)
	    at com.fulcrumgenomics.commons.io.Writer.$plus$plus$eq$(Writer.scala:45)
	    at com.fulcrumgenomics.bam.api.SamWriter.$plus$plus$eq(SamWriter.scala:101)
	    at com.fulcrumgenomics.umi.CallMolecularConsensusReads.execute(CallMolecularConsensusReads.scala:167)
	    at com.fulcrumgenomics.cmdline.FgBioMain.makeItSo(FgBioMain.scala:143)
	    at com.fulcrumgenomics.cmdline.FgBioMain.makeItSoAndExit(FgBioMain.scala:119)
	    at com.fulcrumgenomics.cmdline.FgBioMain$.main(FgBioMain.scala:83)
	    at com.fulcrumgenomics.cmdline.FgBioMain.main(FgBioMain.scala)

I'll have a look into the bam file and see if any of the reads don't match up with the quality scores.

**Comment by jst43** [comment]
@tfenne You were right; I found a single 57bp read with no associated quality score that was introduced at the bwa mem stage. Once I removed it the bam passed CallMolecularConsensusReads without error. Thanks for your help.

**Comment by lbaxter21** [comment]
Hi @jst43, I have this exact same issue - do you remember how you identified the read which was causing the problem? Thank you in advance for any advice!

**Comment by jst43** [comment]
@lbaxter21 I think I either read the bam file into python to look for a malformed read, or I used pysam to catch the read. It was just a quick script so I'm afraid I don't have the code I used.

**Comment by lbaxter21** [comment]
thanks - in the meantime I managed to identify that one of my fastq files had been incompletely downloaded, (using a unix one-liner: for i in `ls *.gz`; do echo $(gzcat ${i} | wc -l)/4|bc; done --- so that was what was causing the outofrange error in my case.I was surprised that bwa hadn't complained about a truncated input file or a mismatch between the number of reads in R1 and R2.  Annoying this slipped through and caused a problem later in the pipeline, but serves as a useful reminder for me to sanity-check input data properly!Thanks for your quick reply,L    On Monday, 24 February 2020, 16:42:25 GMT, Joe Sneath Thompson <notifications@github.com> wrote:  
 
 
@lbaxter21 I think I either read the bam file into python to look for a malformed read, or I used pysam to catch the read. It was just a quick script so I'm afraid I don't have the code I used.
